### PR TITLE
Support incomplete ambiguous dates

### PR DESCRIPTION
**Pull request by huddlej**
### Description of proposed changes    

This PR continues work from #623 by adding support for incomplete ambiguous date strings like those matching the formats of "YYYY" and "YYYY-MM".

### Related issue(s)  

Related to #602.

### Testing

Adds new unit tests for the case of incomplete ambiguous date strings.

### Prior to merge

This PR should be rebased onto master after #623 is merged to avoid a messy git history.

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -102,15 +102,25 @@ def test_read_mask_file_drm_file(self, tmpdir):
         assert utils.read_mask_file(drm_file) == expected_sites
 
     def test_is_date_ambiguous(self):
-        """ is_date_ambiguous should return true for ambiguous dates"""
+        """is_date_ambiguous should return true for ambiguous dates and false for valid dates."""
+        # Test complete date strings with ambiguous values.
         assert utils.is_date_ambiguous("2019-0X-0X", "all")
         assert utils.is_date_ambiguous("2019-XX-09", "month")
         assert utils.is_date_ambiguous("2019-03-XX", "day")
```
Comment (jameshadfield) [inline tests/test_utils.py:109]: Should add the following (will fail tests)
```
        assert utils.is_date_ambiguous("20XX-01-09", "month")
```

File: tests/test_utils.py [hunk 2]
```diff
@@ -102,15 +102,25 @@ def test_read_mask_file_drm_file(self, tmpdir):
...
+        # Test incomplete date strings with ambiguous values.
+        assert utils.is_date_ambiguous("2019", "all")
+        assert utils.is_date_ambiguous("201X", "year")
+        assert utils.is_date_ambiguous("2019-XX", "month")
+        assert utils.is_date_ambiguous("2019-10", "day")
+
+        # Test complete date strings without ambiguous dates for the requested field.
+        assert not utils.is_date_ambiguous("2019-09-03", "all")
+        assert not utils.is_date_ambiguous("2019-03-XX", "month")
+        assert not utils.is_date_ambiguous("2019-09-03", "day")
```
Comment (jameshadfield) [inline tests/test_utils.py:121]: Should consider adding following, even though it passes
```
        assert utils.is_date_ambiguous("2019-XX", "all")
        assert utils.is_date_ambiguous("2019-XX", "day")
```

File: tests/test_utils.py [hunk 3]
```diff
@@ -102,15 +102,25 @@ def test_read_mask_file_drm_file(self, tmpdir):
...
-    def test_not_is_date_ambiguous(self):
-        """ is_date_ambiguous should return false for valid dates"""
-        assert utils.is_date_ambiguous("2019-09-03", "all") is False
-        assert utils.is_date_ambiguous("2019-03-XX", "month") is False
-        assert utils.is_date_ambiguous("2019-XX-01", "day") is False
-        assert utils.is_date_ambiguous("2019-XX-XX", "year") is False
+        # Test incomplete date strings with ambiguous values.
+        assert utils.is_date_ambiguous("2019", "all")
+        assert utils.is_date_ambiguous("201X", "year")
+        assert utils.is_date_ambiguous("2019-XX", "month")
```
Comment (jameshadfield) [inline tests/test_utils.py:115]: Should add following, which currently fail
```
        assert utils.is_date_ambiguous("2019-XX-03", "day")
        assert utils.is_date_ambiguous("20XX-01-03", "day")
```

File: augur/utils.py [hunk 4]
```diff
@@ -84,16 +84,26 @@ def is_date_ambiguous(date, ambiguous_by="all"):
...
+        year, month, day = date_components
+    elif len(date_components) == 2:
+        year, month = date_components
+        day = "XX"
+    else:
+        year = date_components[0]
+        month = "XX"
+        day = "XX"
+
     return (
```
Comment (jameshadfield) [inline augur/utils.py:99]: Is the indentation of the conditionals like this for a reason? I would have expected similar to how the above if/else block is indented (and my linter agrees!)

File: augur/utils.py [hunk 5]
```diff
@@ -84,16 +84,26 @@ def is_date_ambiguous(date, ambiguous_by="all"):
...
-            (ambiguous_by == 'all' and 'X' in date) or
+            (ambiguous_by == 'all' and ('X' in year or 'X' in month or 'X' in day)) or
             (ambiguous_by == 'day' and 'X' in day) or
```
Comment (jameshadfield) [inline augur/utils.py:101]: To go with the tests I added (below), I suggest the logic here is changed to
```
    return (
        (ambiguous_by == 'all' and ('X' in year or 'X' in month or 'X' in day)) or
        (ambiguous_by == 'day' and ('X' in year or 'X' in month or 'X' in day)) or
        (ambiguous_by == 'month' and ('X' in year or 'X' in month)) or
        (ambiguous_by == 'year' and ('X' in year))
    )
```

File: augur/filter.py [hunk 6]
```diff
@@ -106,7 +106,7 @@ def register_arguments(parser):
     parser.add_argument('--include-where', nargs='+',
                                 help="Include samples with these values. ex: host=rat. Multiple values are processed as OR (having any of those specified will be included), not AND. This rule is applied last and ensures any sequences matching these rules will be included.")
     parser.add_argument('--exclude-ambiguous-dates-by', choices=['all', 'day', 'month', 'year'],
-                                help='Exclude ambiguous dates Ex: days - excludes 2020-09-XX, months - excludes 2020-xx-19, all - excludes any ambiguous dates')
+                                help='Exclude ambiguous dates by day (e.g., 2020-09-XX), month (e.g., 2020-XX-XX), year (e.g., 200X-10-01), or all date fields')
```
Comment (jameshadfield) [inline augur/filter.py:109]: For me, `any` is a better word to use than `all`, as the latter implies to me that all fields must be ambiguous.

**Review by jameshadfield — CHANGES_REQUESTED** [review]
Nice work @huddlej - another super useful addition to augur 👍

File: augur/utils.py (hunk 5 above)
Comment (huddlej) [reply augur/utils.py]: Excellent. This helped me realize the redundancy in this original approach to the tests. I've included your logic here in the original code and then refactored it to simplify the logic.

File: augur/utils.py (hunk 4 above)
Comment (huddlej) [reply augur/utils.py]: I missed this in my original code review of #623! :) Thank you for catching it.

**Pull request merged**